File: backend/downloader/utils/browser.py
from playwright.async_api import async_playwright, Page, Browser
from downloader.config import (
    BROWSER_LAUNCH_TIMEOUT,
    NAVIGATION_TIMEOUT,
    ADBLOCK_PATTERNS,
)


async def create_page() -> tuple[Page, Browser]:
    """Crée un navigateur Chromium avec furtivité et blocage pubs."""
    p = await async_playwright().start()
    browser = await p.chromium.launch(
        headless=True,
        args=[
            "--no-sandbox",
            "--disable-blink-features=AutomationControlled",
            "--disable-dev-shm-usage",
            "--disable-gpu",
            "--disable-web-security",
            "--disable-features=IsolateOrigins,site-per-process",
            "--window-size=1920,1080",
        ],
    )
    context = await browser.new_context(
        viewport={"width": 1920, "height": 1080},
        user_agent=(
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/130.0.0.0 Safari/537.36"
        ),
        locale="en-US",
        timezone_id="America/New_York",
        permissions=["notifications"],
        java_script_enabled=True,
    )

    page = await context.new_page()
    page.set_default_timeout(NAVIGATION_TIMEOUT * 1000)

    # ── Bloqueur publicitaire réseau ──────────────────────────
    async def adblock_route(route):
        url = route.request.url.lower()
        for pattern in ADBLOCK_PATTERNS:
            if pattern in url:
                await route.abort()
                return
        await route.continue_()

    await page.route("**/*", adblock_route)

    # ── Évasion de détection automatisation ───────────────────
    await page.add_init_script("""
        Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
        Object.defineProperty(navigator, 'plugins', {
            get: () => [1, 2, 3, 4, 5],
        });
        Object.defineProperty(navigator, 'languages', {
            get: () => ['en-US', 'en'],
        });
        window.chrome = { runtime: {} };
    """)

    return page, browser

File: scrapper/src/config/browser.ts
import { chromium, LaunchOptions, BrowserContext, Page } from 'playwright';

export const browserConfig: LaunchOptions = {
  headless: true,
  executablePath: process.env.PLAYWRIGHT_CHROMIUM_EXECUTABLE_PATH || undefined,
  args: [
    '--no-sandbox',
    '--disable-setuid-sandbox',
    '--disable-dev-shm-usage',
    '--disable-gpu',
    '--disable-blink-features=AutomationControlled'
  ]
};

export const getBrowser = async () => {
  return await chromium.launch(browserConfig);
};

export async function setupFastContext(context: BrowserContext) {
  await context.route('**/*', (route) => {
    const resourceType = route.request().resourceType();
    if (['image', 'media', 'font', 'stylesheet'].includes(resourceType)) {
      return route.abort();
    }
    return route.continue();
  });
}

export async function setupFastPage(page: Page) {
  await page.route('**/*', (route) => {
    const resourceType = route.request().resourceType();
    if (['image', 'media', 'font', 'stylesheet'].includes(resourceType)) {
      return route.abort();
    }
    return route.continue();
  });
}

